repose-aggregator/components/filters/header-id-mapping/src/main/scala/org/openrepose/filters/headeridmapping/HeaderIdMappingFilter.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2016 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+
+package org.openrepose.filters.headeridmapping
+
+import java.net.URL
+import javax.inject.{Inject, Named}
+import javax.servlet._
+import javax.servlet.http.HttpServletRequest
+
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.openrepose.commons.config.manager.UpdateListener
+import org.openrepose.commons.utils.http.PowerApiHeader
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+import org.openrepose.core.filter.FilterConfigHelper
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.filters.headeridmapping.config.{HttpHeader, HeaderIdMappingConfig}
+
+import scala.collection.JavaConverters._
+
+@Named
+class HeaderIdMappingFilter @Inject()(configurationService: ConfigurationService)
+  extends Filter with UpdateListener[HeaderIdMappingConfig] with LazyLogging {
+
+  import HeaderIdMappingFilter._
+
+  private var configurationFile: String = DEFAULT_CONFIG
+  private var initialized = false
+  private var configuredHeaders: Iterable[HttpHeader] = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    logger.trace("Header Identity Mapping filter initializing...")
+    configurationFile = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+
+    logger.info(s"Initializing Header Identity Mapping filter using config $configurationFile")
+    val xsdUrl: URL = getClass.getResource(SCHEMA_FILE_NAME)
+    configurationService.subscribeTo(filterConfig.getFilterName, configurationFile, xsdUrl, this, classOf[HeaderIdMappingConfig])
+
+    logger.trace("Header Identity Mapping filter initialized.")
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    val wrappedRequest = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
```

Comment:
Oh, the init method is absolutely called eagerly, as required by the Filter specification. My concern is over the actual reading of the configuration file. For filters that require configuration, processing (probably) can't occur until that configuration has been presented. That configuration may not be presented when the filter is constructed, however. We subscribe to the ConfigurationService which will notify the filter of configuration changes (well, not deletions, but otherwise). Then the filter sets the "initialized" flag on its listener when the filter has finished processing the configuration. Generally, that is when the filter can actually start processing requests. Prior to that, behavior is undefined.

Requests will not have access to filters that haven't had "init" called on them, as far as I know.